### ISSUE TITLE
SpotBugs 4.8.6: Missing DLS_DEAD_LOCAL_STORE warning for primitive types

### ISSUE THREAD

**Issue by youngchae00** [spotbugs/spotbugs#3100]
I'm trying to understand why SpotBugs doesn't report DLS_DEAD_LOCAL_STORE warnings for primitive types.

I've noticed that when I assign a value to a primitive type variable and then don't use it later in the code, SpotBugs doesn't flag it as a dead store. 

My code is as follows:
```
public void test() {
	int testIntVar = 11;
	String aaa = new String("aaaa"); // report
	Integer testInteger = new Integer(12); // report
	String testString = "13";
	String testString2 = "14";
	byte byteTest = 100;
	short shortTest = 30_000;
	int intTest = 10000;
	long longTest = 100000000;
	float f = 1.456f;
	double d = 1.456789012345678;
	char c = 'c';
	boolean b = true;
	int output = Integer.parseInt(testString2);
	Boolean bt = new Boolean(true); // report
	int ttest = 1+1;
	String bbbb = new String("bbbb"); // report
}
```

![image](https://github.com/user-attachments/assets/627a16c8-9724-4055-a333-cd65c6d8479b)

The image shows that SpotBugs reports warnings for variables `aaa`, `testInteger`, `bt`, and `bbbb`.

I'm wondering if this is expected behavior or if there's a configuration setting I'm missing. Could you please provide some clarification on this?

**Additional information:**

* SpotBugs version: 4.8.6
* Java version: 1.8

Thank you for your time and assistance.

**Comment by gtoison** [comment]
Constant values seem to be specifically ignored when looking for dead stores, I'm not quite sure why TBH:

https://github.com/spotbugs/spotbugs/blob/d5602766acb22de9b04ff817d7e752b79ebcfbf5/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindDeadLocalStores.java#L403

**Comment by JuditKnoll** [comment]
> > Constant values seem to be specifically ignored when looking for dead stores, I'm not quite sure why TBH:
> > https://github.com/spotbugs/spotbugs/blob/d5602766acb22de9b04ff817d7e752b79ebcfbf5/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindDeadLocalStores.java#L403
> 
> This line was added in this [commit](https://github.com/spotbugs/spotbugs/commit/6eaf76c6a796eb2e48d305aaf0470e9a1d65fc56). Without it, it looks like there is an FP in the `hashCode()` of [sfBugs.Bug1864046](https://github.com/spotbugs/spotbugs/blob/master/spotbugsTestCases/src/java/sfBugs/Bug1864046.java) for the variable `prime`. I'm not sure this is the best way to solve this FP, but it's a solution. It looks like the test for this code example is missing currently. I'll check whether I can find other class examples missing the test for them, then open a PR with it.

Added some tests as part of https://github.com/spotbugs/spotbugs/pull/3106 (one for [sfBugs.Bug1864046](https://github.com/spotbugs/spotbugs/blob/master/spotbugsTestCases/src/java/sfBugs/Bug1864046.java) as well).

**Comment by youngchae00** [comment]
https://github.com/spotbugs/spotbugs/blob/d5602766acb22de9b04ff817d7e752b79ebcfbf5/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindDeadLocalStores.java#L404

https://github.com/spotbugs/spotbugs/blob/d5602766acb22de9b04ff817d7e752b79ebcfbf5/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindDeadLocalStores.java#L444

https://github.com/spotbugs/spotbugs/blob/d5602766acb22de9b04ff817d7e752b79ebcfbf5/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindDeadLocalStores.java#L487

I've commented out lines 404, 444, and 487 in FindDeadLocalStores.java.
While this change allows me to detect DLS_DEAD_LOCAL_STORE warnings for primitive types, I'm encountering two false positives.

1. final Variables: The analysis flags final variables as dead code, even though they cannot be reassigned. This is likely due to compiler optimizations that eliminate unnecessary variable loads. For instance, the compiler might directly substitute the constant value at compile time, making it appear as if the variable is unused.

2. Loop Index (Integer): Spotbugs identifies loop iterators declared as Integer (instead of int) as dead code, even though they are essential for loop functionality.

Is there a way to configure FindDeadLocalStores to account for these compiler optimizations and avoid false positives for final variables and loop index variables declared as Integer?

Thank you for your time and assistance.

**Comment by JuditKnoll** [comment]
Which java compiler did you use? Does this arise when using other compilers as well?

If the final variable usage is optimized out in compiler, then I don't think there is an easy way to get rid of this FP. See the note at the bug description:
> Note that Sun's javac compiler often generates dead stores for final local variables. Because SpotBugs is a bytecode-based tool, there is no easy way to eliminate these false positives.
